tsl/src/nodes/gapfill/gapfill_exec.c

```diff
@@ -1380,7 +1380,11 @@ gapfill_exec_expr(GapFillState *state, Expr *expr, bool *isnull)
 	ExprState *exprstate = ExecInitExpr(expr, &state->csstate.ss.ps);
 	ExprContext *exprcontext = GetPerTupleExprContext(state->csstate.ss.ps.state);
 
-	exprcontext->ecxt_scantuple = state->scanslot;
+	/* We can execute expression over a generated tuple which fills the gap (state->scanslot)
+	 * or over an input tuple (state->subslot) to override NULL value when NULLs are treated as
+	 * missing. Use the correct tuple slot for the purpose.
+	 */
+	exprcontext->ecxt_scantuple = (TupIsNull(state->scanslot) ? state->subslot : state->scanslot);
```

Comment:
This is a generic function that is called for calculating many different expressions, not only locf. This logic is locf-specific and not applicable to e.g. evaluating the start/finish arguments. Let's pass the argument tuple table slot as an argument to this function.